### ISSUE TITLE
State feels redundant and obligates too much at the same time

### ISSUE THREAD

**Issue by Deorigosa** [Freshly/flow#46]
Right now we have flow, which implies state. Operations have to use state, whether we want them or not, they can't be called without it. If we want to write to state inside of any operations, we must explicitly name the state option from the very beginning. Also this whole metaprogramming implicit naming (`OneUsefulFlow` and `OneUsefulState`)...

That feels conceptually wrong. I feel like state should exist not for the flow, but rather for operations and their interconnections.

My suggestion is to use [Monads](https://en.wikipedia.org/wiki/Monad_(functional_programming)) (also check out [dry-monads](https://dry-rb.org/gems/dry-monads/1.0/), they did it right but with wonky DSL). This way we can pass around objects, that have outputs/state of previous operation inside, and also explicitly say if the operation was a success or a failure.

**Comment by Deorigosa** [comment]
Also there should be a way to call operations with just a number of keyword args

**Comment by garside** [comment]
Eh, I think I disagree with some of this. Conceptually speaking `OneUsefulFlow` and `OneUsefulState` are not hard requirements, but mere suggestions on a best practice. The whole idea of code reusability is to provide the clarity of the composite structure that exists at a point in time.

Imagine you had three operations you wanted to do: `OperationA, OperationB, OperationC`

OperationA *requires* input `foo`, `bar`, `baz`.
OperationB *requires* `gaz` and optionally takes a `foo`
OperationC *requires* `bar` and internally creates or defines takes a `baz`

In this case, `MyBigFlow` which did:

```ruby
class MyBigFlow
  operations OperationA, OperationB, OperationC
end
```

Would have the following state:

```ruby
class MyBigState
  argument :foo
  argument :bar
  argument :baz
  argument :gaz
end
```

As when these operations are combined together, their interaction makes all input required.

Consider now:

```ruby
class MyMediumFlow
  operations OperationB, OperationC
end
```

this would have a completely different state:

```ruby
class MyMediumState
  argument :gaz
  argument :bar
  option :foo, default: :bananas
  attribute :baz
end
```

Conceptually speaking states are the composite super-state of the required/optional/created input of the consistent operations. Hand crafting a state with a super targeted and segmented interface is a zealous observation of the `I` in solid, the `Interface Separation Principal`. 

If you have data in your state that none of the operations in the flow use... why? What's the benefit of having a bunch of unused data muddying up your understanding of the picture. In the examples above, the redundant obligation of the states is a feature, not a bug IMO.

Otherwise, you have to write even more meta programming about state management to understand that if OperationA requires X and OperationB has an optional X, then X is required in states where the flow will call both, and optional if only one is called. This to me feels like a nice thing for a human developer to bring to the table.

**Comment by garside** [comment]
> Operations have to use state, whether we want them or not, they can't be called without it. 

You can provide a `nil` value to an operation for its state:

```ruby
class AddNumbers < ApplicationOperation
  def behavior
    puts "2 + 2 = #{2+2}"
  end
end
```

```
AddNumbers.new(nil).execute!
{:event=>"execute_started.AddNumbers"}
2 + 2 = 4
{:event=>"execute_finished.AddNumbers"}
=> #<AddNumbers:0x00007f95653b9ed8 @state=nil, @was_executed=true>
```

And per the documentation, they can work with any structured data input:

```ruby
class AddNumbersFancy < ApplicationOperation
  delegate :num1, :num2, to: :state

  def behavior
    puts "#{num1} + #{num2} = #{num1+num2}"
  end
end

AddNumbersFancy.new(OpenStruct.new(num1: 2, num2: 3)).execute
{:event=>"execute_started.AddNumbersFancy"}
2 + 3 = 5
{:event=>"execute_finished.AddNumbersFancy"}
=> #<AddNumbersFancy:0x00007f9564f6f120 @state=#<OpenStruct num1=2, num2=3>, @was_executed=true>
```

**Comment by garside** [comment]
> If we want to write to state inside of any operations, we must explicitly name the state option from the very beginning.

This is a feature not a bug.

If I showed you an object called:

```ruby
class EverythingNeededToDoX
  argument :foo
  argument :bar
  option :baz
end
```

And I asked you "What are all the things needed to do X?" and you told me "foo, bar, baz", how happy would you be if I responded "FALSE! it's foo, bar, baz, and the data it creates, gaz, haz, and snaz which operations ingest into it."

The whole *point* of a state is to require you to actually specify the full and complete shape of the object which you are reading data from and writing data to. 

It is a hardline conceptual feature of flow that the state obligates you to explicitly name the state option from the very beginning. :)

**Comment by garside** [comment]
> Also this whole metaprogramming implicit naming (OneUsefulFlow and OneUsefulState)...

Again, this is just a best practice designed to follow the "principal of least surprise".

You can, if you so desire, do the following:

```ruby
class MakeThingsDoStuff < ApplicationState
  def foo
    :bar
  end

  def bar
    :foo
  end
end

class MakeAThingHappen < ApplicationOperation
  attr_reader :my_state_only

  def initialize(state)
    super(nil)
    @my_state_only = state
  end

  def behavior
    my_state_only.class.attr_accessor :surprise_data
    my_state_only.surprise_data = true
  end
end

class DoSomethingElse < ApplicationOperation
  def behavior
    puts "foo: #{state.foo}"
    puts "bar: #{state.bar}"
    puts "surprise_data: #{state.surprise_data}"
  end
end

class DoAThing < ApplicationFlow
  def self.state_class
    MakeThingsDoStuff
  end

  operations MakeAThingHappen, DoSomethingElse
end
```

And, quite sadly, that all works fine:

```ruby
DoAThing.trigger
{:event=>"trigger_started.DoAThing"}
{:event=>"execute_started.MakeAThingHappen"}
{:event=>"execute_finished.MakeAThingHappen"}
{:event=>"execute_started.DoSomethingElse"}
foo: bar
bar: foo
surprise_data: true
{:event=>"execute_finished.DoSomethingElse"}
{:event=>"trigger_finished.DoAThing"}
=> #<DoAThing:0x00007f9564e14708 @state=#<MakeThingsDoStuff >, @executed_operations=[#<MakeAThingHappen:0x00007f9564dfdf80 @state=nil, @my_state_only=#<MakeThingsDoStuff >, @was_executed=true>, #<DoSomethingElse:0x00007f9564dfdf58 @state=#<MakeThingsDoStuff >, @was_executed=true>], @rewound_operations=[], @memoization_store=#<ShortCircuIt::MemoizationStore memoized: [:operation_instances]>>
```

**Comment by garside** [comment]
> I feel like state should exist not for the flow, but rather for operations and their interconnections.

Per some of the comments above, I would argue that it does. A state is a "frozen" and very explicit summation of the total complexity of a given workflow. Having a document with no surprises that fully documents the voracious complexity of an infinitely extensibly thing is a worthwhile endeavor.

If you disagree that having a single source of truth of what all the information needed for and manipulated by a given complex workflow is a value, then I think we're just an impasse. 

I don't like it when my software surprises me. :(

**Comment by garside** [comment]
> My suggestion is to use Monads (also check out dry-monads, they did it right but with wonky DSL). 

I mean... why?

```ruby
def update_user_address(params)
  maybe_user = Maybe(User.find(params[:user_id]))

  maybe_user.bind do |user|
    maybe_address = Maybe(Address.find(params[:address_id]))

    maybe_address.bind do |address|
      user.update(address_id: address.id)
    end
  end
end
```

Is that really clearer than the rails equivalent:

```ruby
def update_user_address(params)
  user = User.find(params[:user_id])
  user.update!(address: Address.find(params[:address_id]))
rescue ActiveRecord::RecordNotFound
  nil
end
```

These two implementations of `update_user_address` function /identically/. 

I would argue the implementation in the top is inferior to the one below, both syntactically and understandability. 

| Feature | Monads | Operations |
| -------- | ------- | ---------- | 
| pass around objects | ✅ |  ✅ | 
| that have outputs | ✅ |  ✅ | 
| state of previous operation inside  | ✅ |  ✅ | 
| explicitly say if success or a failure | ✅ |  ✅ | 

I guess I don't understand how these kinda oddly shaped monad type object bring anything to the table that's different than an Operation.

Generally speaking, a Flow is an implementation of the command pattern around a state that contains (and only contains) the data that is both necessary and sufficient to perform a complex behavior. Instead of having monadic daisy chains with irreducibly complex of input/output/state changes being just completely discarded. That's basically what our Interactor pattern was/is, and it's overly cumbersome.

I can think of no single use case that a dry-monad approach would bring to the table that would be a superior implementation. But I've been wrong before!

I'll go in on a challenge with you. Pick anything you'd like that you think a dry-monad approach could do better and more clearly, and I'll eat my hat if Flow can't do it better and clearer with less code. :)

**Comment by garside** [comment]
@Deorigosa 

I'll go in on a challenge with you. Pick anything you'd like that you think a dry-monad approach could do better and more clearly, and I'll eat my hat if Flow can't do it better and clearer with less code. :) 

(Moved this comment to it's own so it didn't get lost in the wall of text above)

I've got some free time this weekend if you wanna throw down monad-v-operation. :)

Would be helpful to see how you're thinking about it, cause I'm confident I can make it happen with these operations, but they might need a bit of a massaged DSL depending on the use case.

Excited to see what comes of this! Interesting discussion at least. Hopefully you found my responses good spirited and in defense of the technology! Not trying to be an 🍆

**Comment by garside** [comment]
> Also there should be a way to call operations with just a number of keyword args

Agreed! I think one thing is that there may be some way to touch this up slightly and have everything work. I think realistically you could just do something like this:

```ruby
module Flow
  class OperationBase
    def self.execute!(**input)
      new(OpenStruct.new(input)).execute!
    end
  end
end
```

Then you can really easily call these things in isolation directly outside of a Flow.

Both of the following examples should work: 

```ruby
MyMonasticOperation.execute!(some_data: true, other_data: false)
```

And:

```ruby
class MonasticState
  argument :some_data
  option :other_data, default: false
end

MyMonasticOperation.new(MonasticState.new(some_data: true))
```

**Comment by Deorigosa** [comment]
> And I asked you "What are all the things needed to do X?" and you told me "foo, bar, baz", how happy would you be if I responded "FALSE! it's foo, bar, baz, and the data it creates, gaz, haz, and snaz which operations ingest into it."

I'm completely obsessed with how ActiveInteractions manage inputs. What about this DSL? This way state is completely redundant for the gem user (might be needed under the hood):

```ruby
class SomeFlow
  # raise exception if inputs not provided or invalid
  input :input1
  inputs :input2, :input3, validate_with: InputValidator

  # raise exception if outputs don't match
  output :input1, class: SomeClass
  output :output2
  output :output3, default: false
  output :output4, optional: true

  operations OperationOne, OperationTwo
end

class OperationOne
  inputs :input1, :input2
  
  output :input1, :intermittent

  def behavior
    intermittent = :input2
  end
end

class OperationTwo
  inputs :input1, :intermittent
  
  output :input1, :output2

  def behavior
    output2 = 1
  end
end

# ----------------- #

flow_result = SomeFlow.trigger(input1: :input1, input2: :input2, input3: :input3)
a, b, c, d = flow_result.outputs

# By the way, execute and trigger should be consistent imo
operation_result = OperationOne.execute(input1: :input1, input2: :input2)
e, f = operation_result.outputs
```

**Comment by Deorigosa** [comment]
I'm gonna think a little more about comparison between monads and flows. WIll answer later

**Comment by garside** [comment]
Couple observations on this:

> I'm completely obsessed with how ActiveInteractions manage inputs.

I am less-so. Several things I disagree with about how it works:

1) Typechecking is a severe and intentional violation of the Liskov Substitution Principle. I'm vehemently opposed to it, as I believe it erodes on of the fundamental pillars of SOLID which are extensively battle tested and elegant guidelines for building extensible and maintainable objects. Violating it for essentially the sake feels incorrect to me.

2) Single Responsibility Principle. States manage input. Flows call operations sequentially. Operations do things. Where is the value in throwing away the concept of a precise state object in favor of creating an object with more responsibility? 

3) Not sure if this was an intentional part of the example or an oversight, but it's a point I'm contesting none the less. You have a Flow which defines inputs and outputs which are not used at all in any way by the operations within the flow. `input3` is never used but required. Likewise with `output3` and `output4`. Just extra data bits for variety of input validations or intentional choice?

4) I don't see a particular value in distinguishing output specifically. If I want to output to a controller the source data, I should be able to as well. Seems silly to have to do like `result.output + result.input` to get a total result so I can access both bits of data.

I mean maybe I'm missing something, but this seems conceptually equivalent to what already exists but... with no explicit state class? I guess I just don't agree that having 2 classes handle 3 distinct responsibilities is better than 3 classes doing it.

**Comment by Deorigosa** [comment]
What I'm trying to say is that Flow in its current state doesn't follow duck typing principle. But I'm always bad at saying things properly. Lemme think a bit more to outline it properly

**Comment by Deorigosa** [comment]
As for your answer, I've got a couple counter-arguments:

1) I'm more about the DSL, not the actual strict typing.
2) I'm not against states as a tool under the hood. What I'm saying is that states in their current state don't allow duck typing. More later
3) That was an oversight. And `output3` and `output4` are just for an example. Actually this output pattern really feels bad and also violates duuck typing
4) Ditto

**Comment by Deorigosa** [comment]
```
###############   FLOW   ###############
    |         |         |         |
    |         |  input  |         |
   var1      var2      var3      var4
    |         |         |         └─────────┐
    |         |         └──────────────────┐│
###########   OPERATION1   #############   ││
#   |  input  |                        #   ││
#   0         0 <- used                #   ││
#                                      #   ││  <- not used if not needed
#  var5      var6                      #   ││
#   |  output |                        #   ││
########################################   ││
    |         |         ┌──────────────────┘│
    |         |         |         ┌─────────┘
   var5      var6      var3      var4
    |         |         |         |
    |         |         |         |
###########   OPERATION2   #############
#   |                                  #
#   0                                  #
#                                      ###########
#  var7                                    var8  #
#   |                                       |    #
##################################################
    |         |         |         |         |
    |         |         |         |         |
   var7      var6      var3      var4      var8
    |         |         |         |         |
    |         |         |         |         |
################   FLOW OUTPUT   #################
```

```ruby
class SomeFlow
  operations Operation1, Operation2
end

class Operation1
  inputs :var1, :var2
  outputs :var5, :var6

  validates :var1, presence: true
  validate :some_validation

  def execute
    var5, var6 = :hello, :world
  end

  private

  def some_validation
    true
  end
end

class Operation2
  input :var5
  outputs :var7, :var8

  validates :var1, presence: true
  validate :some_validation

  def execute
    do_something(var5)
    var7, var8 = :hello, :world
  end

  private

  def do_something(*); end
end
```

```ruby
result = SomeFlow.trigger(var1: 1, var2: 2, var3: 3)
puts result.outputs # => { :var3 => 3, :var6 => :world, :var7 => :hello, :var8 => :world }
puts result.outputs.var3 # => 3

op_result = Operation1.execute(var123: 123)
op_result .outputs # => nil
op_result.failure? # => true

# And so on
```

**Comment by garside** [comment]
Oh I see! This has come up a number of different times and ways now. @arettberg is actually working on a solution for this problem. In lieu of `input` and `output`, he's doing I believe `state_accessor` / `state_reader` / `state_writer` which is more idiomatic to the ruby lang I think.

Otherwise the rest of this (sans validation) seems like out-of-the-box behavior.

Also damn bro, respect. Solid ascii-art skills. Color me impressed @Deorigosa :D 🤩

**Comment by garside** [comment]
Wanted to pull a separate comment aside to talk about the validations. I love the idea, but I dunno if it makes sense to put another set of validations in operations as well as states. That to me feels a bit duplicative.

I think instead I would prefer maybe an imperative form of the existing declarative `failure` methods. How would something like this strike you?

```ruby
class Operation1 < ApplicationOperation
  state_reader :weekly_orders

  failure :uneditable_weekly_orders, unless: -> { weekly_orders.all?(&:editable?) }
end
```

Would support the expected `unless:` and `if:` trailing conditionals with symbol support for methods as well.

I just would prefer this rather than having to trudge in the concept of validations again and tie them to a more generic kind of failure. Having states *validate* and operations *fail* because of "bad" states feels like a cleaner division of concerns to me.

Thoughts?

**Comment by garside** [comment]
Also per the `output` thing, I was thinking about this in a totally unrelated concept and came upon the same name. I was thinking of introducing it as the replacement for the half-baked `attribute` accessor that exists right now.

```ruby
class ExampleState < ApplicationState
  argument :foo # something required
  option :bar # something optional
  option :baz, default: :static # with static default
  option(:gaz) { :cool } # with dynamic default

  output :haz # accessor that ONLY exists AFTER initialization. Specifically NOT for input!
  output :bark, default: [] # standard use case for default data values here
  output(:quark) { SecureRandom.hex }
end
```

If you were to do this:

```ruby
ExampleState.new(haz: :something) # => NoMethodError, "haz= is not a method".
```

However, within an operation, you can do this nicely:

```ruby
class Operation1 < ApplicationOperation
  state_writer :haz

  def behavior
    haz = :cheeseburger
  end
end
```

And that'll work great!

Also, love the idea of adding a `.outputs` hash to group all that! Would work nicely with these two combined systems I think!

**Comment by Deorigosa** [comment]
State still makes the whole flow highly coupled, and operations (or flows) can't exist without it. State means too much for an operation, which, in my opinion, should handle all inputs, outputs and behavior. Operations mention state so often that they become feature-envious.

My personal opinion is that all validations must be a trait of the operation, and state should be a supporting object under the hood, rather than an explicit thing that we pass around. Flow should pass state from one operation to another, and handle exceptions gracefully.

I'm gonna iterate another time on the duck-typing. Operations should take inputs and provide outputs. Flow should provide its inputs to operation, collect outputs, provide inputs for the next operation and so on. If any operation can't get desired inputs, it should fail. That's duck typing in this case. Flow should fulfill `input` contract for operations, operations should validate inputs and do things, so flow can fulfill its contract for the next operation. You dislike strict typing, but make things more ugly than that.

**TL; DR:** Operations should be able to exist without state, state should not validate inputs, flows should only orchestrate exceptions and inputs

**Comment by Deorigosa** [comment]
> I mean... why?

I mean `Success` and `Failure` monad objects, not the syntax

> The whole point of a state is to require you to actually specify the full and complete shape of the object which you are reading data from and writing data to.

That violates Ruby's *oh-so-loved* duck typing

> I love the idea, but I dunno if it makes sense to put another set of validations in operations as well as states. That to me feels a bit duplicative.

> Thoughts?

Ditto, but I'd rather put all validations inside operations, as I've said above. DSL does not matter, but I rather `ActiveModel::Validations`

> Also per the output thing, I was thinking about this in a totally unrelated concept and came upon the same name. I was thinking of introducing it as the replacement for the half-baked attribute accessor that exists right now.

For me, state should not handle anything explicitly except an initial input and initial output. **BUT** - the DSL for that should exist on the flow, not state. That will make end-users work only with 2 objects instead of 3, and will make a more explicit representation of what flow should take as inputs and provide as outputs.

**Comment by garside** [comment]
I don't see a benefit in working with 2 objects over three. There feel like three distinct responsibilities here, I don't see the value in shoehorning one of those responsibilities over to another object purely for the sake of eliminating one of the objects because somehow 2 is better than 3?

Here's a conceptual challenge, cause I think this resents one of the best "features" of flow is being able to change the context simply in state without making other alterations.

Mind you, these are actual flows out of Repositron so they're real business cases, not contrived.

Assuming you can wave a magic wand and have whatever DSL "just work", what would the following look like in the system you're suggesting?

=====

## Operations

```ruby
class AssignToWorksheet < ApplicationOperation
  def behavior
    work_units.each { |work_unit| relation_for(worksheet_for(work_unit)) << work_unit }
  end

  protected

  def work_units; end
  def project_for(work_unit); end
  def relation_for(worksheet); end

  private

  def worksheet_for(work_unit)
    employee = work_unit.employee
    timecard = employee.timecards.find_or_create_by! day: Day.for_date(work_unit.posted_at.to_date)
    workbook = employee.workbooks.find_or_create_by! project: project_for(work_unit)

    workbook.worksheets.find_or_create_by! timecard: timecard
  end
end
```

```ruby
class AssignCommitsToWorksheet < ApplicationOperation
  protected

  def work_units
    state.commits
  end

  def relation_for(worksheet)
    worksheet.commits
  end

  def project_for(commit)
    return commit.repository.default_project if commit.pull_requests.none?

    pull_request_for(commit).project
  end

  private

  def pull_request_for(commit)
    @pull_request_for ||= {}
    @pull_request_for[commit] ||= most_relevant_pull_request_for(commit)
  end

  def most_relevant_pull_request_for(commit)
    return if commit.pull_requests.none?
    return commit.pull_requests.first if commit.pull_requests.one?

    valid_pull_requests = commit.pull_requests.select { |pull_request| pull_request.project.reportable? }
    target_pull_requests = valid_pull_requests.any? ? valid_pull_requests : commit.pull_requests
    merged_pull_requests, unmerged_pull_requests = target_pull_requests.partition(&:merged_at?)

    return merged_pull_requests.max_by(&:merged_at) if merged_pull_requests.any?

    unmerged_pull_requests.max_by(&:opened_at)
  end
end
```

```ruby
class AssignCommentsToWorksheet < AssignToWorksheet
  protected

  def work_units
    state.comments
  end

  def relation_for(worksheet)
    worksheet.comments
  end

  def project_for(comment)
    comment.pull_request.project
  end
end
```

## Flows

```ruby
class CalculateWorksheetsFlow < ApplicationFlow
  operations AssignCommitsToWorksheet, AssignCommentsToWorksheet
end
```

```ruby
class RecalculateWorkbooksFlow < CalculateWorkbooksFlow; end
```

## States

```ruby
class CalculateWorkbooksState < ApplicationState
  def workbooks
    @workbooks ||= Workbook.where("updated_at > summarized_at").or(Workbook.where(summarized_at: nil))
  end
end
```

```ruby
class RecalculateWorkbooksState < ApplicationState
  def workbooks
    @workbooks ||= Workbook.all
  end
end
```

**Comment by Deorigosa** [comment]
`AssignCommitsToWorksheet` does not inherit from `AssignToWorksheet`, needfix?

**Comment by garside** [comment]
> State still makes the whole flow highly coupled, and operations (or flows) can't exist without it. 

This is just false. I've shown several examples of how you can work with Operations without even using a state object. True, Operations need an object to take input and produce output. That's a minimum requirement to fulfill the [Command Pattern](https://hackjutsu.com/2015/11/07/Command%20Pattern/) as I understand it.

> State means too much for an operation, which, in my opinion, should handle all inputs, outputs and behavior. Operations mention state so often that they become feature-envious.

Then I think we just disagree. That's just the Interactor pattern we have now where there is no state and Operations "just exist" and have input/output. To answer the question of what I need to work with X Interactors / Organizers, and the answer is basically uncomputable by a machine in the current format.

It just seems like a conceptual wrinkle for you. Operations *already do* handle input, output, and behavior. State is just the mechanism by which things are handed off, rather than having it be ad-hoc.

> My personal opinion is that all validations must be a trait of the operation, and state should be a supporting object under the hood, rather than an explicit thing that we pass around. 

I disagree 100% with this. Consider the following contrived case, and let's assume there is no state, only object validations.

```ruby
class ExampleFlow
  operations Operation1, Operation2
end

class Operation1 < ApplicationOperation
  input :foo 
  input :name

  output :bar 

  validates :name, length: { minimum: 3 }

  def behavior
    foo.update!(name: name)
  end
end

class Operation2 < ApplicationOperation
  input :bar, required: true
  input :foo 
  input :name, length: { maximum: 2 }

  def behavior
    foo.update!(name: "#{name} is cool!")
  end
end

class Operation3 < ApplicationOperation
  option :foo 
  option :bar
  validates :foo, absence: true, if: -> { bar.present? }

  def behavior
    bar.foos << foo
  end
end
```

How would this work?? I have created two mutually incompatible Operations and joined them together in the same flow. There's even several issues here that need to be resolved:

1. How does Operation2 resolve it's presence validation on the output of Operation1 *before* it runs? It's listed as output in Operation1 and required by Operation2, but nothing actually sets the output value in Operation1.
2. How do you communicate that this is a fundamentally broken to the developer? From a high level glance, this all looks ship shape, but there exists no data in the universe that allows this flow to run. 
3. How do you handle the case where the state STARTS valid, but is made INVALID for subsequent operations?

If you are *proactively* validating all state by running each of the validations against the input given to flow, I believe you'll have an incredibly difficult time solving `(1)`.

Likewise, if you are *reactively* validating the state as operation run over it, how do you prevent a flow from executing the first two operations only to fail on the final validation? What mechanism could you use (and where would you put it) to encapsulate that wrinkle?

> Flow should pass state from one operation to another, and handle exceptions gracefully.

That's literally a perfect description of what a Flow does, right now, today. :)

> I'm gonna iterate another time on the duck-typing. Operations should take inputs and provide outputs. 

Sure, just FYI @arettberg has a PR for this that's a WIP which adds convenience delegators to the Operation to indicate what it's pulling out of state.

> Flow should provide its inputs to operation, collect outputs, provide inputs for the next operation and so on. 

As it does today. 

> If any operation can't get desired inputs, it should fail. 

Indeed. Right now, that would fail with a `NoMethodError`. I don't think there's a "more idiomatic" error to indicate "you didn't define a method I wanted to call" MORE than a `NoMethodError`, so raising something else instead of this feels like... more work for the same amount of clarity?

> That's duck typing in this case. 

I don't see how what exists today violates duck typing at all?

> Flow should fulfill input contract for operations, operations should validate inputs and do things, so flow can fulfill its contract for the next operation. 

That is how it works today. I guess I'm just not clear on what part of this you dislike, cause it sounds like you're just describing aspirational features of the software that presently exist?

> You dislike strict typing, but make things more ugly than that.

I don't understand? How do you mean? Can you provide a code example of this? Having a hard time understanding where you see it failing to meet that.

> TL; DR: Operations should be able to exist without state, state should not validate inputs, flows should only orchestrate exceptions and inputs

Not trying to be glib, but as of right now today:

✅ Operations should be able to exist without state
🔴 state should not validate inputs (I disagree 100%; state should be valid before ANY work is done)
✅ flows should only orchestrate exceptions and inputs

Basically all Flows do is hand off input to a state for validation then pass that in sequence to operations.

**Comment by garside** [comment]
@Deorigosa sorry! Typo in the example. It does inherit, I was typing it before I started copy pasting the whole file rather than just the implementation oops! 🙈

**Comment by Deorigosa** [comment]
Aren't there any more typos? Currently the example does not make much sense, sorry :(

**Comment by garside** [comment]
Oh ah yep sorry! My bad. I had two similarly named ones open for the sample and closed the wrong window u_u

**Comment by Deorigosa** [comment]
That's my response:

## Operations

All operations don't change much - except they handle their own inputs cause they need it.
Validations are run inside of operations, because none of them knows if previous ones provide valid data.

```ruby
class AssignToWorksheet < ApplicationOperation

  #
  # Validate all inputs here instead of state
  #

  def behavior
    work_units.each { |work_unit| relation_for(worksheet_for(work_unit)) << work_unit }
  end

  private

  def work_units; end
  def project_for(work_unit); end
  def relation_for(worksheet); end

  def worksheet_for(work_unit)
    employee = work_unit.employee
    timecard = employee.timecards.find_or_create_by! day: Day.for_date(work_unit.posted_at.to_date)
    workbook = employee.workbooks.find_or_create_by! project: project_for(work_unit)

    workbook.worksheets.find_or_create_by! timecard: timecard
  end
end
```

```ruby
class AssignCommitsToWorksheet < AssignToWorksheet
  # Pick your poison - required, optional, input, etc.
  # DSL does not matter much, but it should be consistent for flows and operations
  # Inputs used by operations are consumed, unless they are explicitly set as an output...
  required :commits
  # ...like so:
  #
  # output :commits
  #
  # This way an output will be passed down if it wasn't changed inside operation.
  # If it was - it will have a new value instead.

  #
  # Validate all inputs here instead of state
  #

  private

  def work_units
    # `required` allows using just `commits`
    commits
  end

  def relation_for(worksheet)
    worksheet.commits
  end

  def project_for(commit)
    return commit.repository.default_project if commit.pull_requests.none?

    pull_request_for(commit).project
  end

  def pull_request_for(commit)
    @pull_request_for ||= {}
    @pull_request_for[commit] ||= most_relevant_pull_request_for(commit)
  end

  def most_relevant_pull_request_for(commit)
    return if commit.pull_requests.none?
    return commit.pull_requests.first if commit.pull_requests.one?

    valid_pull_requests = commit.pull_requests.select { |pull_request| pull_request.project.reportable? }
    target_pull_requests = valid_pull_requests.any? ? valid_pull_requests : commit.pull_requests
    merged_pull_requests, unmerged_pull_requests = target_pull_requests.partition(&:merged_at?)

    return merged_pull_requests.max_by(&:merged_at) if merged_pull_requests.any?

    unmerged_pull_requests.max_by(&:opened_at)
  end
end
```

```ruby
class AssignCommentsToWorksheet < AssignToWorksheet
  required :comments

  #
  # Validate all inputs here instead of state
  #

  private

  def work_units
    comments
  end

  def relation_for(worksheet)
    worksheet.comments
  end

  def project_for(comment)
    comment.pull_request.project
  end
end
```

## Flows

Flows handle inputs and outputs for explicity

```ruby
class CalculateWorksheetsFlow < ApplicationFlow
  # pick your poison - required, optional, input, etc.
  # Default lambda should be handled under the hood, as should be the method call
  optional :commits, default: -> { Commit.calculable.where(worksheet_id: nil).eager_load(:repository, pull_requests: :project) }
  optional :comments, default: :default_comments

  # Example:
  #
  # output :bar, :baz
  # 
  # This^ should raise an exception, because operations do not provide these outputs.
  #
  # If output for the flow is not set, all operation outputs should be ignored after completing the flow
  # Might also call out `nil` for explicity, but it is a default value.
  output nil

  operations AssignCommitsToWorksheet, AssignCommentsToWorksheet

  private

  def default_comments
    Comment.calculable.where(worksheet_id: nil).eager_load(pull_request: :project)
  end
end
```

```ruby
class RecalculateWorkbooksFlow < CalculateWorkbooksFlow
  # Same here, but with an override because of inheritance.
  optional :commits, default: :default_commits
  optional :comments, default: -> { Comment.calculable.eager_load(pull_request: :project) }

  private

  def default_commits
    Commit.calculable.eager_load(:repository, pull_requests: :project)
  end
end
```

## States

They don't exist (explicitly)! They are just a RW object, which is initialized with a flow and changed by operations

**Comment by garside** [comment]
okay @Deorigosa updated the example! Sorry for the confusion.

**Comment by Deorigosa** [comment]
I really love what flow is, but I'd rather state to be implicit

**Comment by garside** [comment]
Yea, explicit state is the one of (if not the biggest) demarcation feature between this and interactor pattern. Implicit state is confusing, brittle, and hard to manage.

I say that from experience. We have it in the current codebase and it's a huge pain in the ass to assume how and where things are defined. We went so far as to do the this style stuff:

```ruby
optional :comments, default: -> { Comment.calculable.eager_load(pull_request: :project) }
```

Specifically thought the constructors, and that's the area of the most confusion in that pattern. I don't view it as "having to create a central state object", to me it's "getting to create a central state object".

It just feels like a better observation of the single responsibility principle. Having flows be responsible for input/output, sequence handling, and operation ordering just felt like too many unrelated responsibilities.

I think we're just at an impasse and disagree with the nature of explicit state. To me, it's the feature that makes Flow more maintainable than Interactors. Without explicit state, we might as well just use the existing interactor pattern, as that's what it already does.

**Comment by garside** [comment]
I think the only other observation here I can say is that I fundamentally believe there are two separate completely unrelated tasks being talked about here that keep getting conceptually bundled together:

* State *validation*
* Operation *sanity*

State validation happens *before any operations are run*, and is a full and complete reckoning of all the good and proper guard clauses a developer can conjure up for a given workflow. 

If the validation passes, it is an indication that to the best of the understanding of the developers, this flow should generate no errors being executed.

Operation sanity happens *as each operation is run*, and has an opportunity to raise a different kind of issue to indicate the break.

I feel there's value in these two things because they indicate two very different problems:

Problem 1: Before doing anything, I have some invalid data. I probably passed wrong input or my expectations on the state of those input were invalid, so the flow will not run.
Problem 2: After validating that the state was correct, an operation failed because it was asked to do something illegal.

In the case of Problem 1, I know to focus my attention on the "top" of the flow, whatever was passing data down into it, because the problem exists outside the universe the flow is creating.

Conversely, Problem 2, I know to focus my attention on the operations. Given I've followed the other rules of the system the only way to get an "illegal operation" error should be that one of my operations earlier in the flow did something which made the CURRENT state being given to the operation invalid, despite the fact that the INITIAL state was well and good.

Trying to distinguish "this is wrong because you told me something bad" vs "this is wrong because I did something unexpected" where validations just live entirely in the one object of the Operation feels impossibly complicated.

Per the example above, you'd need some way of validating "okay you need X, but X isn't going to be provided, but is likely going to be defined by an earlier operation, incase it's not". This will always be a (problem 2) case.

And honestly, if you can find a way to move that specific example into a (problem 1) case, then I do believe you have solved `P!=NP` problem and can go on to fame and fortune as a lecturer in the compute science circuit and likely be memorized forever. :)

Cause I can't think of a way to do that WITHOUT running the operations and seeing what happens.

**Comment by Deorigosa** [comment]
Something along these lines?
```ruby
def check_input_integrity
  operation_classes.each do |operation_class|
    if inputs & operation_class.required_inputs == operation_class.required_inputs
      inputs.delete!(operation_class.inputs).push(*operation_class.outputs)
    else
      raise ArgumentError
  end
end

**Comment by Deorigosa** [comment]
Yet I think that all validations should exist on the operation, not state. But that's a holy way discussion, let's skip that

**Comment by garside** [comment]
it's basically all personal preference I think. I'm working on, better or worse, the proactive failure method so you can "validate" by raising an error. 

I'm not necessarily considering this fundamentally different, I just didn't want two models involved and it felt easy enough to fail that way for now.

**Comment by garside** [comment]
Eh that solution feels a bit blunt in terms of precision. I think it gets you some of the way, but I wouldn't want to rely on it, cause it seems like it could yield some false positives.

I would rather double the relatively minimal effort to summarize a "final state" with all the associated validations for the time being. I do honestly believe you need both to be "complete" and I do feel they are separate things. I don't know I have a clearer way to express my feeling than:

```ruby
class State1 < ApplicationState
  argument :weekly_order

  def delivery
    weekly_order.deliveries.first
  end
end

class State2 < ApplicationState 
  argument :user 
  output :delivery

  def weekly_order
    user.weekly_orders.can_not_deliver_this_week.upcoming.first
  end
end

class Flow1 < ApplicationFlow
  operations EmailDeliveryUpdate
end

class Flow2 < ApplicationFlow
  operations RescheduleDelivery, EmailDeliveryUpdate
end

class RescheduleDelivery < ApplicationOperation
  state_reader :weekly_order
  state_writer :delivery
end

class EmailDeliveryUpdate < ApplicationOperation
  state_reader :delivery
end
```

This feels like a clean division of responsibilities to me. 

In the same way you can use an Operation without a State actually, the inverse is also true. You can use a State in isolation as well!

```ruby
class SomeController < ApplicationController
  def index
    @weekly_orders = upcoming_weekly_orders_state.weekly_orders
  end

  private

  def upcoming_weekly_orders_state
    UpcomingWeeklyOrderState.new(user: current_user)
  end
  memoize :upcoming_weekly_orders_state
end
```

This allows your states to be pretty effortless replacements for "query objects" and can give you a single source of truth for your business logic AND your queries to utilize the same subset of business logic.

That kind of extensibility is ultimately why I see value in an explicit state. I would say every flow should have a state, BUT not every state necessarily needs a flow. At their core, states are just a nice little framework to build a PORO object with some discreetly named input/output channels.

**Comment by Deorigosa** [comment]
Closed as the discussion died
